# Part 1: Data Cleaning

Let's start with getting the datafiles rounds.csv and companies.txt.


In [112]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

# reading data files
# using encoding = "ISO-8859-1" to avoid pandas encoding error
rounds = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
companies = pd.read_csv("companies.txt", sep="\t", encoding = "ISO-8859-1")


In [113]:
# Look at rounds head
print(rounds.head())

                     company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008      

In [114]:
# inspect the structure etc.
print(rounds.info(), "\n")
print(rounds.shape)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 3.1+ MB
None 

(114949, 6)


The variables funding_round_code and raised_amount_usd contain some missing values, as shown above. We'll deal with them shortly.

In [115]:
# look at companies head
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [116]:
# companies structure
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 2.5+ MB


Ideally, the ```permalink``` column in the companies df should be the unique_key of the table, having 66368 unique company names (links, or permalinks). Also, these 66368 companies should be present in the rounds file.

Let's first confirm that these 66368 permalinks (which are the URL paths of companies' websites) are not repeating in the column.

In [117]:
# identify the unique number of permalinks in companies
len(companies.permalink.unique())

66368

Also, let's convert all the entries to lowercase (or uppercase) for uniformity.

In [118]:
# converting all permalinks to lowercase
companies['permalink'] = companies['permalink'].str.lower()
companies.head()


permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [119]:
# look at unique values again
len(companies.permalink.unique())

66368

Thus, there are 66368 unique companies in the table and ```permalink``` is the unique primary key. Each row represents a unique company.

Let's now check whether all of these 66368 companies are present in the rounds file, and if some extra ones are present.

In [120]:
# look at unique company names in rounds df
# note that the column name in rounds file is different (company_permalink)
len(rounds.company_permalink.unique())


90247

There seem to be 90247 unique values of ```company_permalink```, whereas we expected only 66368. May be this is because of uppercase/lowercase issues.

Let's convert the column to lowercase and look at unique values again.

In [121]:
# converting column to lowercase
rounds['company_permalink'] = rounds['company_permalink'].str.lower()
rounds.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [122]:
# Look at unique values again
len(rounds.company_permalink.unique())

66370

There seem to be 2 extra permalinks in the rounds file which are not present in the companies file. Let's hope that this is a data quality issue, since if this were genuine, we have two companies whose investment round details are available but their metadata (company name, sector etc.) is not available in the companies table.

Let's have a look at these two company permalinks.

In [123]:
# companies present in rounds file but not in companies file
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  \
29597                             /organization/e-cãŠbica   
31863        /organization/energystone-games-灵石游戏   
45176                /organization/huizuche-com-惠ç§ÿ车   
58473              /organization/magnet-tech-磁石科技   
101036  /organization/tipcat-interactive-沙èˆÿ信息ç...   
109969             /organization/weiche-tech-喂车科技   
113839                 /organization/zengame-ç¦
游科技   

                                funding_round_permalink funding_round_type  \
29597   /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863   /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176   /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473   /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   
101036  /funding-round/41005928a1439cb2d706a43cb661f60f               seed   
109969  /funding-round/f74e457f838b81fa0b29649740f186d8            venture   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf               seed   

       funding_round_code   funded_at  raised_amount_usd  
29597                 NaN  01-02-2015                NaN  
31863                 NaN  09-08-2014                NaN  
45176                 NaN  18-09-2014                NaN  
58473                 NaN  16-08-2014          1625585.0  
101036                NaN  06-09-2010                NaN  
109969                  A  06-09-2015                NaN  
113839                NaN  17-07-2010                NaN

All the permalinks have weird non-English characters. Let's see whether these characters are present in the original df as well. 

In [124]:
# looking at the indices with weird characters
rounds_original = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
rounds_original.iloc[[29597, 31863, 45176], :]

company_permalink  \
29597                       /ORGANIZATION/E-CÊBICA   
31863  /ORGANIZATION/ENERGYSTONE-GAMES-ǁµÇŸ³Æ¸¸Æˆ   
45176          /organization/huizuche-com-惠ç§ÿ车   

                               funding_round_permalink funding_round_type  \
29597  /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863  /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176  /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   

      funding_round_code   funded_at  raised_amount_usd  
29597                NaN  01-02-2015                NaN  
31863                NaN  09-08-2014                NaN  
45176                NaN  18-09-2014                NaN

The company weird characters appear when you import the data file. To confirm whether these characters are actually present in the given data or whether python has introduced them while importing, let's have a look at the original CSV file in Excel.

The figure below shows the filtered rows - they have normal English characters.

<img src="./weird_names.PNG">

Thus, this is most likely a data quality issue we have introduced while reading (specfically, encoding) the data file into python. 

In [126]:
import chardet

rawdata = open('rounds2.csv', 'rb').read()
result = chardet.detect(rawdata)
charenc = result['encoding']
print(charenc)

Windows-1254


In [145]:
print(result)

{'encoding': 'Windows-1254', 'language': 'Turkish', 'confidence': 0.4186155476629225}


In [159]:
# trying different encodings
rounds_original = pd.read_csv("rounds2.csv", encoding = 'utf32')
rounds_original.iloc[[29597, 31863, 45176], :]

UnicodeDecodeError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)